### PR TITLE
DRAFT  ink! tutorial 

### PR DESCRIPTION
**Pull request by PierreOssun**

### CODE REVIEW

File: docs/builder-guides/from- zero-to-ink-hero/dex/Factory/factory.md [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+    use uniswap_v2::{
+        impls::factory::*,
+        traits::factory::*,
+    };
+```
+
+Add the [storage struct](https://use.ink/macros-attributes/storage) and add the factory field (that we defined in traits):
+
+```rust
+    #[ink(storage)]
```
Comment (bobo-k2) [inline docs/builder-guides/from- zero-to-ink-hero/dex/Factory/factory.md:270]: Remove tab

**Review by bobo-k2 — CHANGES_REQUESTED** [review]
Great work. I went through the tutorial and found and commented some points that could be improved.

File: docs/builder-guides/from- zero-to-ink-hero/dex/Pair/mint.md [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+    let now = Self::env().block_timestamp();
+    let time_elapsed = now - self.data::<data::Data>().block_timestamp_last;
+    if time_elapsed > 0 && reserve_0 != 0 && reserve_1 != 0 {
+        let price_cumulative_last_0 = (reserve_1 / reserve_0)
+            .checked_mul(time_elapsed as u128)
+            .ok_or(PairError::MulOverFlow4)?;
+        let price_cumulative_last_1 = (reserve_0 / reserve_1)
+            .checked_mul(time_elapsed as u128)
+            .ok_or(PairError::MulOverFlow4)?;
+        self.data::<data::Data>().price_0_cumulative_last += price_cumulative_last_0;
```
[the remark this answers is not in the document]
Comment (PierreOssun) [reply docs/builder-guides/from- zero-to-ink-hero/dex/Pair/mint.md:187]: By design =)
The UniswapV2Pair cumulative price variables are designed to eventually overflow, i.e. price0CumulativeLast and price1CumulativeLast and blockTimestampLast will overflow through 0.
https://docs.uniswap.org/contracts/v2/guides/smart-contract-integration/building-an-oracle

File: docs/builder-guides/from- zero-to-ink-hero/dex/uniswap-v2-dex-in-ink.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+Please follow these tutorials first:
+
+| Tutorial                                                                   | Difficulty                     |
+|----------------------------------------------------------------------------|--------------------------------|
+| [Your first flipper contract](../flipper-contract/flipper.md)              | Basic ink! -  Basic Rust       |          
+|  [full dApp: NFT contract with PSP34 + UI](../flipper-contract/flipper.md) | Intermediate ink! - Basic Rust |
+
+# To follow this tutorial you will need:
+- to [set up your ink! environment](../../XVM%20and%20WASM/setup_your_ink_environment.md)
+- to have knowledge about AMM & [Uniswap V2 implementation](https://docs.uniswap.org/contracts/v2/overview) (as this tutorial will fucos on implementation)
```
Comment (Maar-io) [inline docs/builder-guides/from- zero-to-ink-hero/dex/uniswap-v2-dex-in-ink.md:17]: fucos :)

**Comment by Maar-io** [comment]
chapters before block code are sometimes ending with `:` and sometimes without it. Can you make it consistent?

**Comment by Maar-io** [comment]
Can you add some light into why `data::<MyType>()` is called as function

**Comment by PierreOssun** [comment]
> Can you add some light into why data::<MyType>() is called as function
Explanation is here:
https://github.com/Supercolony-net/openbrush-contracts/blob/e366f6ff1e5892c6a624833dd337a6da16a06baa/lang/src/traits.rs#L63

I think this is really advanced Rust and will not bring value for a tutorial. But advanced developer will check the Trait anyway if they need to understand

**Review by fiexer — APPROVED** [review]

**Review by bobo-k2 — APPROVED** [review]
LGTM

**Pull request merged**